Fix source_app/destination_app to respect app.kubernetes.io/name label (#58436)

aviralgarg05 opened the pull request:
**Please provide a description of this PR:**
This change ensures that source_app, destination_app, source_version, and destination_version metric labels properly respect Kubernetes well-known labels (app.kubernetes.io/name and app.kubernetes.io/version) with the correct priority order:

For app name:
1. service.istio.io/canonical-name
2. app.kubernetes.io/name
3. app

For version:
1. service.istio.io/canonical-revision
2. app.kubernetes.io/version
3. version

The fix adds a GetAppName function to the labels package that uses the existing label priority lookup mechanism, and updates serviceClusterOrDefault in bootstrap/config.go to use this function instead of directly accessing the 'app' label.

This maintains backward compatibility while allowing users who only have app.kubernetes.io/name or app.kubernetes.io/version labels to have their metric labels properly populated.

Fixes #58436

aviralgarg05 committed:
Fix source_app/destination_app to respect app.kubernetes.io/name label (#58436)

This change ensures that source_app, destination_app, source_version, and
destination_version metric labels properly respect Kubernetes well-known
labels (app.kubernetes.io/name and app.kubernetes.io/version) with the
correct priority order:

For app name:
1. service.istio.io/canonical-name
2. app.kubernetes.io/name
3. app

For version:
1. service.istio.io/canonical-revision
2. app.kubernetes.io/version
3. version

The fix adds a GetAppName function to the labels package that uses the
existing label priority lookup mechanism, and updates serviceClusterOrDefault
in bootstrap/config.go to use this function instead of directly accessing
the 'app' label.

This maintains backward compatibility while allowing users who only have
app.kubernetes.io/name or app.kubernetes.io/version labels to have their
metric labels properly populated.

Fixes #58436

Stevenjin8 commented:
/ok-to-test

aviralgarg05 committed:
Address PR review feedback: backward-compatible label priority order

Changes from previous commit:
1. Reverted the order of checks to match original behavior: labels are
   checked BEFORE WorkloadName for backward compatibility
2. Added GetAppNameBackwardCompatible function that checks labels in
   order: app -> app.kubernetes.io/name -> canonical-name
3. Updated tests to reflect correct priority order

The change maintains exact backward compatibility:
- If 'app' label exists, it takes priority (same as before)
- If only 'app.kubernetes.io/name' exists, it is now used (new feature)
- If only canonical-name exists, it is used as last fallback (new feature)
- WorkloadName is only used if none of the above labels are present

This ensures existing users see no behavior change while enabling users
who only have app.kubernetes.io/name to have their metrics properly labeled.

sridhargaddam reviewed (commented):
LGTM, thanks @aviralgarg05.
Can you add a release-note so that users are aware of this.

aviralgarg05 committed:
Add release note for issue #58436

Adds release note explaining the new support for app.kubernetes.io/name
and service.istio.io/canonical-name labels when populating source_app
and destination_app metric labels.

aviralgarg05 commented:
Please merge

Stevenjin8 reviewed (commented):
GitHub isn't letting me do inline comments, but one of the function comments had the priority of labels backwards. Otherwise lgtm

Stevenjin8 reviewed (commented):
Sorry, I left these in an uncommitted state

Stevenjin8 reviewed (changes requested):
see comments

aviralgarg05 committed:
Address PR feedback: remove unused GetAppName and rename GetAppNameBackwardCompatible to GetApp

dikderoy commented:
Hi there,

FYI, you probably should include this in the changelog:

This one also affects trace metrics. 

we previously never pushed `app` label, only had `app.kubernetes.io/name` configured, and our trace metrics now shifted `service` dimention after update.

before `service` dimension on trace metrics had `<raw-service-name>.<service-namespace>`

now - `service` dimension has `<whatever is in app.kubernetes.io/name>.<service-namespace>`

which breaks things if one uses `app.kubernetes.io/name` as a common name across all objects in a helm chart, e.g. has multiple services defined, all sharing the said label with the same value.

so now all services are reporting the same metric, while before they were different.

I am also not sure (not familiar with code base) but it does affect the config structure in envoy, or not?

not sure if that matters for networking, likely not, or I haven't noticed it affecting me, but still worth mentioning.


Tbh, this change likely has a bigger impact on the community.

All public helm charts use app.kubernetes.io/name to define application name uniformly across all resources they manage (this is a general recommendation from both kubernetes.io and helm.sh official documentation)

So all helm charts with more than one service will have their traces/trace-metrics collapsed into a single one w/o the ability to get traffic split by an individual service object (edit: there is, via resource_name, but we also used `service` as a selector b/c those metrics are collapsing with gateways w/o it)

While this is fixable on the user side, it's somewhat of a breaking change.
